casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -70,113 +56,209 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
-      validSender  <- Validate.blockSenderHasWeight(b, genesis, dag)
-      validVersion <- Validate.version(b, version)
-      attemptResult <- if (!validFormat) (BlockStatus.invalidFormat.asLeft[ValidBlock], dag).pure
-                      else if (!validSig)
-                        (BlockStatus.invalidSignature.asLeft[ValidBlock], dag).pure
-                      else if (!validSender)
-                        (BlockStatus.invalidSender.asLeft[ValidBlock], dag).pure
-                      else if (!validVersion)
-                        (BlockStatus.invalidVersion.asLeft[ValidBlock], dag).pure
-                      else attemptAdd(b, dag)
-      (attempt, updatedDag) = attemptResult
-      _ <- attempt match {
-            case Left(InvalidBlock.MissingBlocks) => noop
-            case _ =>
-              Cell[F, CasperState].modify { s =>
-                s.copy(
-                  blockBuffer = s.blockBuffer - b.blockHash,
-                  dependencyDag = DoublyLinkedDagOperations.remove(s.dependencyDag, b.blockHash)
-                )
-              }
-          }
-      _ <- Span[F].mark("attempt-result")
-      _ <- attempt match {
-            case Left(ib: InvalidBlock) if !InvalidBlock.isSlashable(ib) => noop
-            case Left(bs) if !BlockStatus.isInDag(bs)                    => noop
-            case _ =>
-              reAttemptBuffer(updatedDag) // reAttempt for any status that resulted in the adding of the block into the view
-          }
-      _         <- Span[F].mark("reattempted-buffer")
-      tipHashes <- estimator(updatedDag)
-      _         <- Span[F].mark("after-estimator")
-      tipHash   = tipHashes.head
-      _         <- Log[F].info(s"New fork-choice tip is block ${PrettyPrinter.buildString(tipHash)}.")
-    } yield attempt
-
-  def contains(
-      hash: BlockHash
-  ): F[Boolean] =
+  /**
+    * Check if there are blocks in CasperBuffer available with all dependencies met.
+    * @return First from the set of available blocks
+    */
+  private def getNextReadyBlock: F[Option[BlockHash]] = {
+    import cats.instances.list._
     for {
-      dag            <- blockDag
-      dagContains    <- dag.contains(hash)
-      state          <- Cell[F, CasperState].read
-      bufferContains = state.blockBuffer.contains(hash)
-    } yield (dagContains || bufferContains)
+      dag      <- blockDag
+      pendants <- CasperBufferStorage[F].getPendants
+      depFreePendants <- pendants.toList.filterA { pendant =>
+                          for {
+                            pendantBlock   <- BlockStore[F].get(pendant)
+                            justifications = pendantBlock.get.justifications
+                            // If even one of justifications is not in DAG - block is not dependency free
+                            depFree <- justifications.foldLeftM(true)(
+                                        (r, j) => r.pure[F] &&^ dag.contains(j.latestBlockHash)
+                                      )
```

Comment:
You could use `justifications.existsM` to just fold until first justification is not found.